app/controllers/api/v0/raw_controller.rb

```diff
@@ -0,0 +1,11 @@
+class Api::V0::RawController < Api::V0::ApiController
+
+  def show
+    @version ||= Version.find_by(version_hash: params[:id])
+    storage_path = Rails.root.join 'tmp/storage/archive'
+    path = "#{storage_path}/#{@version.version_hash}"
+    mime_type = @version.source_metadata['mime_type']
+    send_file(path, type: mime_type, disposition: 'inline')
```

Comment:
So… unless I’m really missing something, this is straight-up not going to work in production or any environment that is not using the `LocalFile` Filestore implementation. It feels like a lot of implementation details are leaking out of `FileStorage` and into the rest of the app here.

This is especially true if we ever decide we need a different kind of storage than S3. What if, for example, we added an IPFS for Qri FileStorage? (Not that I expect to do that in the near future.) I think retrieving the file’s content (or a stream of the file’s content) probably needs to remain the responsibility of the FileStorage implementation (with the nice-but-not-required performance optimization of redirecting to the actual storage location if it’s publicly accessible, like our S3 buckets).

Reply:
Also, keep in mind that the storage path is meant to be configured per instance of `LocalFile`, so this is also assuming something (the path being `Rails.root.join('tmp/storage/archive')`) that should probably be determined at configuration time (not compilation time as here). *If nothing else* we should be reading this info from `Archiver.store` instead of hardcoding it here.

Reply:
> So… unless I’m really missing something, this is straight-up not going to work in production or any environment that is not using the LocalFile Filestore implementation.

Nope, not missing anything. My only implementation goal was to serve files that we have, and 4xx errors for those that we don't.

> I think retrieving the file’s content (or a stream of the file’s content) probably needs to remain the responsibility of the FileStorage implementation (with the nice-but-not-required performance optimization of redirecting to the actual storage location if it’s publicly accessible, like our S3 buckets).

I see how I missed something important here on the first swing. It sounds like the goal is to always be able to always serve up the raw body (if it's accessible to our server), even if it means proxying or bridging protocols. We might just choose for it to redirect instead for efficiency. I think I can do that using aws buckets as my non-local reference implementation.

> Also, keep in mind that the storage path is meant to be configured per instance of `LocalFile`, so this is also assuming something (the path being `Rails.root.join('tmp/storage/archive')`) that should probably be determined at configuration time (not compilation time as here). _If nothing else_ we should be reading this info from `Archiver.store` instead of hardcoding it here (but, as noted in the above comment, I’m not sure that’s really good enough either).

Yeah it looks like I dropped a TODO on not hardcoding that path.

Reply:
> It sounds like the goal is to always be able to always serve up the raw body (if it's accessible to our server), even if it means proxying or bridging protocols.

Yes! Sorry, I didn’t realize that was not clear. From my perspective, the *problem at hand* is not anything to do with storing files locally — it’s that the stored files might not be publicly accessible (for example, you’d have the same problem with if you used `FileStorage::S3` with a private bucket). Any version record we provide metadata for is one we also *have* to provide the raw body for.

> We might just choose for it to redirect instead for efficiency.

Exactly 👍